feat(autoscaling): add `availabilityZoneDistribution` property to an AutoScalingGroup

ren-yamanashi opened the pull request:
### Issue # (if applicable)

n/A

### Reason for this change

We can set a `availabilityZoneDistribution` for an AutoScalingGroup from cloudformation, but this was not supported in the AWS CDK L2 construct.

### Description of changes

Add `availabilityZoneDistribution` property to CommonAutoScalingGroupProps and set it in the CfnAutoScalingGroup.

### Description of how you validated changes

Added both unit and integration tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 'autoscaling' instead of 'aws-autoscaling'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 'autoscaling' instead of 'aws-autoscaling'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 'autoscaling' instead of 'aws-autoscaling'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ec5bc2ac0667eda3e3671aec9a3a417ca3776311
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ea355f74-a0cd-4a4f-bf52-6d282779ad24%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 'autoscaling' instead of 'aws-autoscaling'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 'autoscaling' instead of 'aws-autoscaling'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 0cb2e0823d4e1ee34deb8b80d30958446d7ad40d
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4d6387c9-5cdc-4d7a-96fb-14a1c2bcdb66%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 41b597b817e7c50cb84ee479a1861f68a3a36664
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4717c7e4-154e-4de5-86f7-2dd65962ff3b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b51ccc39553cc1a764fd87e0c4a78db8692b7c5d
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=dabd15d8-cbfd-4369-8d08-b1cf5c39766d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 39f4400b99f0fd5683534a4d21bd76c2593171c7
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7c44627a-dc72-40c4-9ee9-82fef3e23576%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 103c2b253f73b0cc94e05d9b3ca97be9b272aafa
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a3e7c8c4-5c89-4d6e-9f97-7686f673b379%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d84afbf2b272134cbe19ce6f312bea8f6adb0c17
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc0e6c86-1b7c-48ba-aa0a-9b3d35924f69%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ce8ede9d2da1679d5c57ce44e512ce15bcd8959d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6782070e-6724-4dbf-80bc-ec8b0808f25f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d036a0698d9012541751aeb971a8454b3c0162d0
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=82fc62f3-89f4-4589-84bf-02895930446a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6d7773a0e10454407a297168d18d72edc238dafc
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=adc8ffa9-e0f0-4614-a8dc-625555660013%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

ren-yamanashi commented:
> Thank you for the contribution. I've commented on property design. If you have any thoughts on that please let me know.

@mazyu36 

Thank you for your review!  
I have corrected the issue in the following commit, please confirm.

https://github.com/aws/aws-cdk/pull/32100/commits/478f2ac903a51daea043568b7f45606219894f42

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 478f2ac903a51daea043568b7f45606219894f42
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6cbc42e9-ca6b-43e1-9fc8-4dc68a7862fd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

ren-yamanashi commented:
@mergifyio update

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8c8423bb4240cea166fba318d749bc0b54396671
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=84810ced-e816-48bf-a770-a2b8f4c24e36%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

mazyu36 commented on packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:
```diff
@@ -411,6 +411,12 @@ export interface CommonAutoScalingGroupProps {
+
+  /**
+   * The strategy for distributing instances across Availability Zones.
+   * @default BALANCED_BEST_EFFORT
```
```suggestion
   * @default CapacityDistributionStrategy.BALANCED_BEST_EFFORT
```

ren-yamanashi replied:
https://github.com/aws/aws-cdk/pull/32100/commits/385396e97fa674852e9f541f8a99ffdd2e58b1e5

mazyu36 commented on packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:
```diff
@@ -1098,6 +1104,20 @@ export class GroupMetric {
+  /**
+   * If launches fail in an Availability Zone, Auto Scaling will attempt to launch in another healthy Availability Zone instead.
+   */
+  BALANCED_BEST_EFFORT= 'balanced-best-effort',
```
nit: space
```suggestion
  BALANCED_BEST_EFFORT = 'balanced-best-effort',
```

mazyu36 reviewed (commented):
Thanks! Some minor comments.
I'm not sure why Codecov failed.
Let's see if the issue persists after the changes are applied.

ren-yamanashi commented:
> Thanks! Some minor comments. I'm not sure why Codecov failed. Let's see if the issue persists after the changes are applied.

@mazyu36 

Thank you for your comment!  
I have corrected the issue, please confirm.  
( I have included a link to the corrected commit in the thread)

Also, I fixed the default value of `capacityDistributionStrategy` was not applied correctly.
Please check this as well 🙏 

https://github.com/aws/aws-cdk/pull/32100/commits/bda5050b77800d46635684a55e914354ff590770

(Codecov seems to have fixed it too! Thank you 😆 )

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: bda5050b77800d46635684a55e914354ff590770
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e4e659d1-2e5b-4baa-8f47-fa98aa2adc5a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a14ccadd42611e50df332eae7435a4276ea34386
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9ce78c39-bcca-4e68-8d44-a66e94bdbf3c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

mazyu36 commented on packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:
```diff
@@ -1530,6 +1550,9 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
       autoScalingGroupName: this.physicalName,
+      availabilityZoneDistribution: props.azCapacityDistributionStrategy
+        ? { capacityDistributionStrategy: props.azCapacityDistributionStrategy }
+        : { capacityDistributionStrategy: CapacityDistributionStrategy.BALANCED_BEST_EFFORT },
```
Thank you for the fix.
But I think default value should be `undefined`.
The `@default` should describe the behavior at that time.

When the default value is explicitly set, it causes a breaking change.
This is because before and after this PR merged, the synth result changes when `props.azCapacityDistributionStrategy` is omitted.

ren-yamanashi replied:
@mazyu36 

thank you!

I see, I understand.

So, I will also omit the description of default in JSDoc, is that correct?  
(Because there was no mention of default values in the CloudFormation documentation...)  
https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-availabilityzonedistribution.html

mazyu36 replied:
For optional properties, `@default` is necessary. 

Also, it's common that default value behaviors are not documented in the CFn documentation.
In such cases, I often verify the default value by deploying without specifying the value.

When I tried deploying it without `availabilityZoneDistribution` property, the value became `Balanced best effort`, so I think this is the default value.

![image](https://github.com/user-attachments/assets/d6cedcb7-2e62-4d74-9580-af3871860a8d)

ren-yamanashi replied:
Thank you for the details.  
I have corrected it with the following commit!

https://github.com/aws/aws-cdk/pull/32100/commits/ea1e756446da8768dc4cd70b5f885811315b9ed8

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ea1e756446da8768dc4cd70b5f885811315b9ed8
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3135db99-e7d2-46c0-ad9c-4bce6b40312d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

mazyu36 reviewed (approved):
Thanks! LGTM

moelasmar commented on packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:
```diff
@@ -411,6 +411,12 @@ export interface CommonAutoScalingGroupProps {
+
+  /**
+   * The strategy for distributing instances across Availability Zones.
+   * @default CapacityDistributionStrategy.BALANCED_BEST_EFFORT
```
this is the service default value, and not CDK default value. We should not mention the service default value here. I prefer mention that default is undefined

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 0d535e96aab4d3be1f45bf41b528ac7ceccce59a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=60f22a6f-b9da-41f2-8922-6376a915f6e6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ba905751d5cb60379afe25ab6801ddbee8e555ed
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f9b1de07-25f7-4758-8748-3d79f7f9ff50%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
